test/unit/controllers/todo-item.controller.test.ts

```diff
@@ -0,0 +1,260 @@
+      }
+      // Repository stub should not have been called!
+      expect(create.called).to.be.false;
+    });
+  });
+  describe('getTodoItem', () => {
+    it('returns a todo item if it exists', async () => {
+      const controller = new TodoItemController(todoItemRepo, todoRepo);
+      const findTodoItem = todoItemRepo.findById as sinon.SinonStub;
+      const findTodo = todoRepo.findById as sinon.SinonStub;
+      findTodoItem.withArgs(payload.id as number).resolves(payloadWithId);
+      findTodo.withArgs(todo.id as number).resolves(todo);
+      expect(
+        await controller.getTodoItem(todo.id as number, payload.id as number)
+      ).to.eql(payloadWithId);
+    });
+
+    it('throws if it does not exist', async () => {
+      const controller = new TodoItemController(todoItemRepo, todoRepo);
+      const findTodoItem = todoItemRepo.findById as sinon.SinonStub;
+      const findTodo = todoRepo.findById as sinon.SinonStub;
+      findTodoItem.withArgs(payload.id as number).resolves(null);
+      findTodo.withArgs(todo.id as number).resolves(todo);
+      try {
+        await controller.getTodoItem(todo.id as number, payload.id as number);
+      } catch (err) {
+        expect(err).to.match(/No todo item with id /);
+      }
+    });
+  });
+  describe('getTodoItems', () => {
+    it('returns all todo items on a todo', async () => {
+      const controller = new TodoItemController(todoItemRepo, todoRepo);
+      const findTodoItems = todoItemRepo.find as sinon.SinonStub;
+      const findTodo = todoRepo.findById as sinon.SinonStub;
+      const expected = [
+        payloadWithId,
+        { title: 'do another thing', todoId: 2 }
+      ];
+      findTodoItems.resolves(expected);
+      findTodo.withArgs(todo.id as number).resolves(todo);
+      expect(await controller.getTodoItems(todo.id as number)).to.eql(expected);
+    });
+    it('returns an empty list if there are none', async () => {
+      const controller = new TodoItemController(todoItemRepo, todoRepo);
+      const findTodoItems = todoItemRepo.find as sinon.SinonStub;
+      const findTodo = todoRepo.findById as sinon.SinonStub;
+      const expected: TodoItem[] = [];
+      findTodoItems.resolves(expected);
+      findTodo.withArgs(todo.id as number).resolves(todo);
+      expect(await controller.getTodoItems(todo.id as number)).to.eql(expected);
+    });
+  });
+  describe('replaceTodoItem', () => {
+    it('replaces a todo item if it exists', async () => {
+      const controller = new TodoItemController(todoItemRepo, todoRepo);
+      const replace = todoItemRepo.deleteById as sinon.SinonStub;
+      const findTodoItem = todoItemRepo.findById as sinon.SinonStub;
+      const item = Object.assign({}, payload, { id: 1, todoId: 1 });
+      const newItem = Object.assign({}, payload, { title: 'title two' });
+      replace.withArgs(item.id as number, newItem).resolves(newItem);
+      findTodoItem.withArgs(item.id as number).resolves(todo);
```

Comment:
In my experience, this style where the stubs have strict argument matching produces test failures that are difficult to troubleshoot.

For example, if the controller invokes `replace` with id `42` instead of `item.id`, then `replace` returns undefined and sometime later you will see an error that `undefined` was supposed to equal to something else.

Another problem with this approach is that it breaks arrange-act-assert pattern - the assertion that `replace` should be invoked with `item.id` is implicitly hidden in the arrange part where we are setting up the stubbed behavior.

A better approach is to make stubs loose, let them always return the good value regardless of the input arguments. Then add assertions to verify that stubbed method were invoked as expected.

```ts
// arrange
replace.resolves(newItem);

// act
// controller.replaceTodoItem

// assert
sinon.assert.calledWith(replace, item.id as number, newItem);
```

This produces a more helpful test failure:

```
AssertError: expected stub to be called with arguments
1 2
// 1 is printed in green, 2 is printed in red
```

The good approach is described in http://loopback.io/doc/en/lb4/Testing-your-application.html#create-a-stub-repository, but the section does not say why other approaches are bad.  See https://github.com/strongloop/loopback-next/issues/684